src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2259,91 @@ private void EndEntry()
+        /// </summary>
+        private void RestoreOmittedNullValues()
+        {
+            // Restore omitted null value only when processing response and when
+            // Preference header omit-values=nulls is specified.
+            if (this.jsonLightInputContext.ReadingResponse
+                && this.jsonLightInputContext.MessageReaderSettings.NullValuesOmitted)
+            {
+                IODataJsonLightReaderResourceState resourceState = this.CurrentResourceState;
+                IEdmStructuredType edmStructuredType = resourceState.ResourceType;
+
+                if (resourceState.SelectedProperties == SelectedPropertiesNode.Empty)
+                {
+                    return;
+                }
+                else
+                {
+                    if (resourceState.Resource != null && edmStructuredType != null)
+                    {
+                        ODataResourceBase resource = resourceState.Resource;
+
+                        IEnumerable<IEdmProperty> selectedProperties;
+                        if (resourceState.SelectedProperties == SelectedPropertiesNode.EntireSubtree)
+                        {
+                            selectedProperties = edmStructuredType.DeclaredProperties;
+                        }
+                        else
+                        { // Partial subtree. Combine navigation properties and selected properties at the node with distinct.
+                            selectedProperties =
+                                resourceState.SelectedProperties.GetSelectedNavigationProperties(edmStructuredType)
+                                as IEnumerable<IEdmProperty>;
+                            selectedProperties = selectedProperties.Concat(
+                                resourceState.SelectedProperties.GetSelectedProperties(edmStructuredType).Values)
+                                .Distinct();
+                        }
+
+                        foreach (IEdmProperty currentProperty in selectedProperties)
+                        {
+                            Debug.Assert(currentProperty.Type != null, "currentProperty.Type != null");
+                            if (!currentProperty.Type.IsNullable)
+                            {
+                                // Skip declared properties that are not null-able types.
+                                continue;
+                            }
+
+                            // Response should be generated using case sensitive, matching EDM defined by metadata.
+                            // In order to find potential omitted properties, need to check
+                            // resource properties read and navigation properties read.
+                            if (!resource.Properties.Any(
+                                p => p.Name.Equals(currentProperty.Name, StringComparison.Ordinal))
+                                && !resourceState.NavigationPropertiesRead.Contains(currentProperty.Name))

```

Comment:
Make sure we have test case to cover complex type and navigation property.

Reply:
Already have coverage for complex type.
Added coverage for navigation link in e2e test ExpandedCustomerEntryOmitNullValuesTest.
Added various test case for expanded navigation property in PropertyAndValueJsonLightReaderIntegrationTests: expanded navigation link should have no content when the link is null (see sample link: [http://services.odata.org/V4/TripPinService/People](http://services.odata.org/V4/TripPinService/People)('ronaldmundy')?$expand=Photo)

---
In reply to: [173012135](https://github.com/OData/odata.net/pull/1106#discussion_r173012135) [](ancestors = 173012135)

Reply:
I don't think TripPinService is handling this case correctly; if an expanded single-value navigation property is null, I would expect a null value be returned in the payload, and I would expect to get the null value as nested content when I ready the payload.  We should make sure that ODL does support reading and writing a single-value null navigation property, and fix this case to handle a missing explicitly selected null navigation property the same as an explicit null value for a nullable single-value navigation property in the payload.

---
In reply to: [174645486](https://github.com/OData/odata.net/pull/1106#discussion_r174645486) [](ancestors = 174645486,173012135)

Reply:
Will look into this issue about expanded nav property and it should be nestedInfo with null restored.

---
In reply to: [186565783](https://github.com/OData/odata.net/pull/1106#discussion_r186565783) [](ancestors = 186565783,174645486,173012135)